src/proxy.c

```diff
@@ -238,113 +239,131 @@ cc_proxy_msg_read (GIOChannel *source, GIOCondition condition,
 	struct proxy_msg p_msg = { 0 };
 	gchar iov_buffer[LINE_MAX] = { 0 };
 	ssize_t bytes_read;
-	gboolean fd_read = false;
 	struct msghdr msg = { 0 };
 	struct cmsghdr *cmsg = NULL;
 	struct iovec io = { .iov_base = iov_buffer,
 	                    .iov_len = sizeof(iov_buffer) };
-	char ctl_buffer[LINE_MAX] = { 0 };
+	char ctl_buffer[CMSG_SPACE(sizeof(int))] = { 0 };
 	unsigned char *data;
 	/* length + reserved */
 	const size_t data_offset = sizeof(p_msg.length) + sizeof(p_msg.reserved);
 	guint8 buf_len[sizeof(p_msg.length)] = { 0 };
-	size_t bytes_to_copy = 0;
+	unsigned char *frame = NULL;
+	unsigned char *tmp_ptr;
+	int timeout;
+	size_t tot_b_rd = 0;
+	size_t tot_b_expected;
+	ssize_t b_to_cpy;
 
 	msg.msg_iov = &io;
 	msg.msg_iovlen = 1;
 	msg.msg_control = ctl_buffer;
 	msg.msg_controllen = sizeof(ctl_buffer);
 
-	/* read length */
-	bytes_read = recvmsg(proxy_data->socket_fd, &msg, 0);
-
-	if (bytes_read == -1){
-		g_critical("failed to read proxy message %s", strerror(errno));
-		return false;
+	/* 
+	 * This sleep is very important to make sure we don't
+	 * get MSG_CTRUNC message from msghdr. There is no way
+	 * to recover from a MSG_CTRUNC the control message lost.
+	 */
+	if (proxy_data->oob_fd) {
+		usleep(1e3);
 	}
 
-	if (bytes_read == 0) {
-		g_critical("failed to read lenght");
-		return false;
-	}
+	/* read at least message length */
+	for (timeout = 0; timeout < 100; timeout++) {
+		if (tot_b_rd >= sizeof(buf_len)) {
+			break;
+		}
 
-	if (proxy_data->oob_fd) {
-		/* check if oob data was received */
-		cmsg = CMSG_FIRSTHDR(&msg);
-		if (cmsg) {
-			data = CMSG_DATA(cmsg);
-			if (data) {
-				*proxy_data->oob_fd = *((int*) data);
-				g_message("oob fd read from proxy socket %d",
-					*proxy_data->oob_fd);
-				fd_read = true;
-			}
+		bytes_read = recvmsg(proxy_data->socket_fd, &msg, 0);
+		if (bytes_read < 1) {
+			usleep(1e3);
```

Comment:
is this sleep needed only to get oob data? if this is the case you should use it only when oob is expected
```if (bytes_read < 1 && proxy_data->oob_fd) { ```